### PR TITLE
adding events app and Event model

### LINKED ISSUE #26: Create events app and Event model
- create events app and register it in the settings.py file
- add Event model in events/models.py
- add tests for Event model in events/tests.py

### PR DESCRIPTION
**Pull request by BelikovYoav**
- add events app files 
- update settings.py to register events app
- add Event model in events/models.py
- add tests for Event model in events/tests.py

fix: #26

### CODE REVIEW

**Review by MortarDefender — APPROVED** [review]
Looks great, changes are approved

**Review by YaelD — APPROVED** [review]
Looks great

**Comment by kobihk** [comment]
@MortarDefender @YaelD 
could you please look at the comments me and @nemerna added in:
https://github.com/redhat-beyond/lets-meet/pull/29
and try to add the same in this PR (where needed)
Thanks in advance,
 @kobihk

**Comment by MortarDefender** [comment]
@kobihk yes, we have to review your comments on Yael PR, and we will have an update today.
We will take what is relevant from your comments on Yael PR and will change it in this PR as well

File: events/tests.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        except ValidationError as e:
+            assert 'start and end date_time - one field not filled' in e.message_dict[NON_FIELD_ERRORS]
+
+    def test_start_bigger_equal_end_event(self, new_event):
+        try:
+            new_event.date_time_start, new_event.date_time_end = DATE_TIME_LATER, DATE_TIME_EARLY
+            new_event.save()
+            assert False, ERROR
+        except ValidationError as e:
+            assert 'date_time_start is equal or greater than date_time_end' in e.message_dict[NON_FIELD_ERRORS]
```
Comment (MortarDefender) [inline events/tests.py]: Maybe it will be even better to use the parametrize as there are a few functions that are essentially the same,
You can use this link to Yael's PR as a reference :
https://github.com/redhat-beyond/lets-meet/blob/4a43276a865c15a024686ff3fec53217c1563b5e/users/tests.py

In addition, a website that can explain it much better:
https://realpython.com/pytest-python-testing/#parametrization-combining-tests

File: events/models.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        BROWN = "#A52A2A", "Brown"
+        WHITE = "#FFFFFF", "White"
+        GRAY = "#808080", "Gray"
+        BLACK = "#000000", "Black"
+
+    title = models.CharField(max_length=50)
+    location = models.CharField(blank=True, max_length=50)
+    description = models.TextField(blank=True)
+    date_time_start = models.DateTimeField(null=True, blank=True)
+    date_time_end = models.DateTimeField(null=True, blank=True)
```
Comment (YaelD) [inline events/models.py]: I don't think that we should provied the start date and end date fields to be blank or null. 
In my opinion these fields should be mandatory.

File: events/migrations/0001_initial.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    initial = True
+
+    dependencies = [
+    ]
+
```
Comment (MortarDefender) [inline events/migrations/0001_initial.py:12]: Please add some test data, as you can see in Yael's PR linked below
https://github.com/redhat-beyond/lets-meet/blob/4a43276a865c15a024686ff3fec53217c1563b5e/users/migrations/0003_user_test_data.py

In addition, another example can be seen at 
https://drive.google.com/file/d/1BM_P51iCnlRs9OVxTx2woxjfO-FoukRr/view
pages 44 - 45

**Review by MortarDefender — CHANGES_REQUESTED** [review]

File: events/models.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+           or (self.date_time_start is None and self.date_time_end is not None)):
+            raise ValidationError('start and end date_time - one field not filled')
+
+        if (self.date_time_start is not None and self.date_time_end is not None
+           and self.date_time_start >= self.date_time_end):
+            raise ValidationError('date_time_start is equal or greater than date_time_end')
+
+    def save(self, *args, **kwargs):
+        self.full_clean()
+        return super().save(*args, **kwargs)
```
Comment (YaelD) [inline events/models.py]: Maybe using a pre-save signal can be better- without ovrriding an existing functions.
Here is an explanation: https://docs.djangoproject.com/en/4.0/topics/signals/
You can also see an example of the useage in signals in my PR: 
https://github.com/redhat-beyond/lets-meet/blob/4a43276a865c15a024686ff3fec53217c1563b5e/users/models.py
(lines 64-69)

**Review by YaelD — CHANGES_REQUESTED** [review]

File: events/models.py (hunk 2 above)
Comment (BelikovYoav) [reply events/models.py]: done

File: events/tests.py (hunk 1 above)
Comment (BelikovYoav) [reply events/tests.py]: done, added parametrize tests functions

File: events/migrations/0001_initial.py (hunk 3 above)
Comment (BelikovYoav) [reply events/migrations/0001_initial.py:12]: done, added some test data

File: events/models.py (hunk 4 above)
Comment (BelikovYoav) [reply events/models.py]: done, change to signals

**Review by YaelD — DISMISSED** [review]
Looks good. 
Greate job!

**Review by MortarDefender — DISMISSED** [review]
Looks great

**Review by nemerna-finastra — DISMISSED** [review]

**Review by nemerna — COMMENTED** [review]
Looks Good
@brochwerger , may be you can make an extra scan, i have not seen any special wrong modifications

**Comment by RonHachmon** [comment]
looks great, approved

**Review by RonHachmon — DISMISSED** [review]
looks great, approved

**Comment by MortarDefender** [comment]
@kobihk @brochwerger
Can you review this PR, we have made some changes last night according to your comments.

**Review by kobihk — CHANGES_REQUESTED** [review]
@BelikovYoav 
I can see that you have the same files we already merged in: #25 
could you please make sure your local branch is rebased on the latest code?
or maybe I missed something...

**Comment by BelikovYoav** [comment]
> @BelikovYoav I can see that you have the same files we already merged in: #25 could you please make sure your local branch is rebased on the latest code? or maybe I missed something...

@kobihk 
changed as you requested

File: events/models.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        choices=Colors.choices,
+        default=Colors.BLACK,
+    )
+
+    def __str__(self) -> str:
+        return f"{self.title} - {self.date_time_start}"
+
+
+def validate_blank_title(title):
+    if title is None:
```
Comment (kobihk) [inline events/models.py]:
```suggestion
    if not title:
```
if the "title" is equal to an empty string the validation will pass

as you can see:
In [3]: a = ""

In [4]: if a is None:
   ...:     print("a is None")
   ...: else:
   ...:     print("a is NOT None")
   ...: 
a is NOT None

File: events/models.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        raise ValidationError('title cannot be blank')
+
+
+def validate_date_time(date_time_start, date_time_end):
+    if date_time_start is None:
```
Comment (kobihk) [inline events/models.py]:
```suggestion
    if not date_time_start:
```

File: events/models.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        raise ValidationError('Starting date cannot be blank')
+    if date_time_end is None:
```
Comment (kobihk) [inline events/models.py]:
```suggestion
    if not date_time_end:
```

File: events/models.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        raise ValidationError('Ending date cannot be blank')
+    if date_time_start >= date_time_end:
+        raise ValidationError(f'{date_time_start} must be smaller than {date_time_end}')
+
+
+def event_fields_validation(sender, instance, **kwargs):
```
Comment (kobihk) [inline events/models.py]: AFAICS the sender and kwargs not in use 
maybe you can get rid of them

Comment (kobihk) [inline events/models.py]: maybe you can change the "instance" to something less generic 
```suggestion
def event_fields_validation(sender, current_event, **kwargs):
```

File: events/tests.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
+from django.core.exceptions import ValidationError
+from django.utils import timezone
+from .models import Event
+from datetime import datetime
+
+import pytest
+
+
+DATE_TIME_EARLY = datetime(2022, 3, 24, 12, 12, 12, 0, tzinfo=timezone.utc)
```
Comment (kobihk) [inline events/tests.py]: maybe it will be more readable like:
```suggestion
DATE_TIME_START = datetime(2022, 3, 24, 12, 12, 12, 0, tzinfo=timezone.utc)
```

File: events/tests.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+DATE_TIME_LATER = datetime(2022, 3, 24, 14, 12, 12, 0, tzinfo=timezone.utc)
```
Comment (kobihk) [inline events/tests.py]:
```suggestion
DATE_TIME_END = datetime(2022, 3, 24, 14, 12, 12, 0, tzinfo=timezone.utc)
```

File: events/tests.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+TITLE = 'new_title'
+
+
+@pytest.fixture
+def new_event():
+    return Event(title='new_title', location='new_location', description='new_description',
```
Comment (kobihk) [inline events/tests.py]: please use the constant from line # 11
and create the same for location and description
```suggestion
    return Event(title=TITLE, location='new_location', description='new_description',
```

File: events/tests.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+
+@pytest.fixture
+def persist_event(db, new_event):
+    new_event.save()
+    return new_event
+
+
+@pytest.mark.django_db
+def test_persist_event(persist_event):
```
Comment (kobihk) [inline events/tests.py]: maybe now that you have test data you can use it here instead of create new one

File: events/tests.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    assert persist_event not in Event.objects.all()
+
+
+def create_event(title, date_time_start, date_time_end):
+    return Event(title=title, date_time_start=date_time_start, date_time_end=date_time_end)
+
+
+@pytest.mark.parametrize('events, excpected_error', [
+    (create_event(None, DATE_TIME_EARLY, DATE_TIME_LATER), 'title cannot be blank'),
+    (create_event(TITLE, None, DATE_TIME_LATER), 'Starting date cannot be blank'),
```
Comment (kobihk) [inline events/tests.py]: Please add tests with an empty string

File: events/tests.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    persist_event.delete()
+    assert persist_event not in Event.objects.all()
+
+
+def create_event(title, date_time_start, date_time_end):
+    return Event(title=title, date_time_start=date_time_start, date_time_end=date_time_end)
+
+
+@pytest.mark.parametrize('events, excpected_error', [
+    (create_event(None, DATE_TIME_EARLY, DATE_TIME_LATER), 'title cannot be blank'),
```
Comment (kobihk) [inline events/tests.py]: please prevent repeating yourself like calling the create_event in every line instead of doing it once inside the method 
```suggestion
    (None, DATE_TIME_EARLY, DATE_TIME_LATER, 'title cannot be blank'),
```

File: events/tests.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    (create_event(TITLE, None, DATE_TIME_LATER), 'Starting date cannot be blank'),
+    (create_event(TITLE, DATE_TIME_EARLY, None), 'Ending date cannot be blank'),
+    (create_event(TITLE, DATE_TIME_EARLY, DATE_TIME_EARLY),
+     f'{DATE_TIME_EARLY} must be smaller than {DATE_TIME_EARLY}'),
+    (create_event(TITLE, DATE_TIME_LATER, DATE_TIME_EARLY), f'{DATE_TIME_LATER} must be smaller than {DATE_TIME_EARLY}')
+])
+def test_invalidation(events, excpected_error):
```
Comment (kobihk) [inline events/tests.py]: maybe something like:
```suggestion
def test_invalid_valuses(title, date_time_start, date_time_end, excpected_error):
```

File: events/tests.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    try:
+        events.save()
+    except ValidationError as error:
+        assert excpected_error in error  # The event was not created successfully
+    else:
+        assert False  # The event was created successfully
```
Comment (kobihk) [inline events/tests.py]: maybe something like:
```suggestion
    current_error = ''
    try:
        events.save()
    except ValidationError as error:
        current_error = error
    assert excpected_error == current_error
```

**Review by kobihk — CHANGES_REQUESTED** [review]

File: events/models.py (hunk 5 above)
Comment (BelikovYoav) [reply events/models.py]: done, changed it as suggested and now validation raises error

File: events/models.py (hunk 6 above)
Comment (BelikovYoav) [reply events/models.py]: done

File: events/models.py (hunk 7 above)
Comment (BelikovYoav) [reply events/models.py]: done

File: events/models.py (hunk 8 above)
Comment (BelikovYoav) [reply events/models.py]: deleted signals and now using override save function, admin page had problems with ValidationError signals

Comment (BelikovYoav) [reply events/models.py]: deleted signals and now using override save function, admin page had problems with ValidationError signals

File: events/tests.py (hunk 9 above)
Comment (BelikovYoav) [reply events/tests.py]: done

File: events/tests.py (hunk 10 above)
Comment (BelikovYoav) [reply events/tests.py]: done

File: events/tests.py (hunk 11 above)
Comment (BelikovYoav) [reply events/tests.py]: done

File: events/tests.py (hunk 12 above)
Comment (BelikovYoav) [reply events/tests.py]: - this test is to check that we can add new event and that the new event is saved as expected, changed test name to test_add_event for better understanding.

- also added new test called test_exist_event that checks events in test data exists

File: events/tests.py (hunk 13 above)
Comment (BelikovYoav) [reply events/tests.py]: done

File: events/tests.py (hunk 14 above)
Comment (BelikovYoav) [reply events/tests.py]: done

File: events/tests.py (hunk 15 above)
Comment (BelikovYoav) [reply events/tests.py]: done

File: events/tests.py (hunk 16 above)
Comment (BelikovYoav) [reply events/tests.py]: done

**Review by MortarDefender — DISMISSED** [review]

File: events/models.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+            raise ValidationError('title cannot be blank')
+        if not self.date_time_start:
+            raise ValidationError('Starting date cannot be blank')
+        if not self.date_time_end:
+            raise ValidationError('Ending date cannot be blank')
+        if self.date_time_start >= self.date_time_end:
+            raise ValidationError(f'{self.date_time_start} must be smaller than {self.date_time_end}')
+
+    def save(self, *args, **kwargs):
+        self.clean()
```
Comment (kobihk) [inline events/models.py:42]: AFAIU this is a validation method so maybe you can call it something like:
```suggestion
        self.validate_params()
```

File: events/tests.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+DATE_TIME_END = datetime(2022, 3, 24, 14, 12, 12, 0, tzinfo=timezone.utc)
+TITLE = 'new_title'
+LOCATION = 'new_location'
+DESCRIPTION = 'new_description'
+
+
+@pytest.fixture
+def new_event():
+    return Event(title=TITLE, location=LOCATION, description=DESCRIPTION,
+                 date_time_start=DATE_TIME_START, date_time_end=DATE_TIME_END)
```
Comment (kobihk) [inline events/tests.py]: The convention is to open the parentheses and create a new line then add what you need and close the parenthesis in a new line - in this way it is more visible where it start and end
```suggestion
    return Event(
        title=TITLE, location=LOCATION, description=DESCRIPTION,
        date_time_start=DATE_TIME_START, date_time_end=DATE_TIME_END
    )
```

**Review by kobihk — CHANGES_REQUESTED** [review]

File: events/tests.py (hunk 18 above)
Comment (BelikovYoav) [reply events/tests.py]: done

File: events/models.py (hunk 17 above)
Comment (BelikovYoav) [reply events/models.py:42]: Screenshot_3

changing the methods name breaks the admin page, from what i understood from django documentation
django expects validation to be inside a clean method

**Review by MortarDefender — APPROVED** [review]

**Review by RavidYael — APPROVED** [review]

File: events/models.py (hunk 17 above)
Comment (kobihk) [reply events/models.py:42]: @BelikovYoav please take a look at the error in the "Exception Value"
btw I download your PR, changed it locally, run vagrant up... and add an event without any problem

Comment (BelikovYoav) [reply events/models.py:42]: @kobihk the problem occurs when you try to add an event with date time start bigger than date time end, not when you try to add a "good" event

Comment (kobihk) [reply events/models.py:42]: @BelikovYoav I see 
Thanks!!

**Review by kobihk — APPROVED** [review]
